crates/sui-graphql-rpc/src/types/move_package.rs

```diff
@@ -575,6 +629,75 @@ impl Loader<PackageVersionKey> for Db {
     }
 }
 
+#[async_trait::async_trait]
+impl Loader<LatestKey> for Db {
+    type Value = SuiAddress;
+    type Error = Error;
+
+    async fn load(&self, keys: &[LatestKey]) -> Result<HashMap<LatestKey, SuiAddress>, Error> {
+        use packages::dsl;
+        let other = diesel::alias!(packages as other);
+
+        let mut ids_by_cursor: BTreeMap<_, BTreeSet<_>> = BTreeMap::new();
+        for key in keys {
+            ids_by_cursor
+                .entry(key.checkpoint_viewed_at)
+                .or_default()
+                .insert(key.address.into_vec());
+        }
+
+        // Issue concurrent reads for each group of IDs
+        let futures = ids_by_cursor
+            .into_iter()
+            .map(|(checkpoint_viewed_at, ids)| {
+                self.execute(move |conn| {
+                    let results: Vec<(Vec<u8>, Vec<u8>)> = conn.results(|| {
```

Comment:
not unique to this particular implementation, but these concurrent reads in our dataloaders may exhaust the database connection pool; it's possible for someone then to cook up some request that will hog up all connections. By default dataloader is configured to coalesce 1k individual loads, so I don't think it'll be too hard to figure out a query that would exceed the pool size configured for our graphql service

Reply:
This is a good call-out -- it's worth batching these so we only issue 10 reads at a time per data loader (and I think it's also worth upping the default connection limit to like 50 or something).

I'll do this in a follow-up PR that handles all the cases where we do this kind of grouping.

Reply:
Looked into this. We only ever send out parallel requests when we group requests that could have distinct checkpoint viewed at cursors. This means that we're not introducing more connection parallelism than is already present in the underlying query (the DataLoader sends out at most as many parallel requests as an unoptimised query would).

This means that even if we limited the fan-out within data-loader enabled queries, someone could create the same issue with some other kind of query, which suggests that the right thing to do would be to apply a more generic limit (which we already do -- the input node limit) and tweak that to preserve the connection pool.